File: 2025/day02/perl/solution.pl
#!/usr/bin/env perl
use strict;
use warnings;

# Read input file
my $input_file = '../input.txt';
open(my $fh, '<', $input_file) or die "Cannot open $input_file: $!";
my $input = do { local $/; <$fh> };
close($fh);
chomp $input;

# Check if a number is invalid (pattern repeated exactly twice)
sub is_invalid_id {
    my ($num) = @_;
    my $str = "$num";
    my $len = length($str);

    # Must be even length to be repeated twice
    return 0 if $len % 2 != 0;

    my $half_len = $len / 2;
    my $first_half = substr($str, 0, $half_len);
    my $second_half = substr($str, $half_len);

    # Check if both halves are identical
    return $first_half eq $second_half;
}

# Check if a number is invalid (pattern repeated at least twice)
sub is_invalid_id_part2 {
    my ($num) = @_;
    my $str = "$num";
    my $len = length($str);

    # Try all possible pattern lengths from 1 to len/2
    for (my $pattern_len = 1; $pattern_len <= $len / 2; $pattern_len++) {
        # Check if the string can be divided evenly by this pattern length
        if ($len % $pattern_len == 0) {
            my $pattern = substr($str, 0, $pattern_len);
            my $repetitions = $len / $pattern_len;

            # Check if repeating the pattern creates the full string
            if ($repetitions >= 2 && $pattern x $repetitions eq $str) {
                return 1;
            }
        }
    }

    return 0;
}

# Part 1: Sum of invalid IDs where pattern repeated exactly twice
sub part1 {
    my ($input) = @_;

    # Parse the comma-separated ranges
    my @range_strs = split(/,/, $input);
    my @ranges = map {
        my ($start, $end) = split(/-/, $_);
        { start => $start, end => $end }
    } @range_strs;

    my $sum = 0;

    # Check each range for invalid IDs
    foreach my $range (@ranges) {
        for (my $id = $range->{start}; $id <= $range->{end}; $id++) {
            if (is_invalid_id($id)) {
                $sum += $id;
            }
        }
    }

    return $sum;
}

# Part 2: Sum of invalid IDs where pattern repeated at least twice
sub part2 {
    my ($input) = @_;

    # Parse the comma-separated ranges
    my @range_strs = split(/,/, $input);
    my @ranges = map {
        my ($start, $end) = split(/-/, $_);
        { start => $start, end => $end }
    } @range_strs;

    my $sum = 0;

    # Check each range for invalid IDs
    foreach my $range (@ranges) {
        for (my $id = $range->{start}; $id <= $range->{end}; $id++) {
            if (is_invalid_id_part2($id)) {
                $sum += $id;
            }
        }
    }

    return $sum;
}

# Run both parts
print "Part 1: ", part1($input), "\n";
print "Part 2: ", part2($input), "\n";

File: 2025/day02/java/Solution.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

public class Solution {
    private static String inputText;

    public static void main(String[] args) throws IOException {
        Path inputPath = Path.of(args.length > 0 ? args[0] : "../input.txt");
        inputText = Files.readString(inputPath).trim();

        System.out.println("Part 1: " + part1());
        System.out.println("Part 2: " + part2());
    }

    private static boolean isInvalidIdPart1(long num) {
        String s = Long.toString(num);
        int length = s.length();

        // Must have even length to be repeated twice
        if (length % 2 != 0) {
            return false;
        }

        // Check if it starts with 0 (leading zeros not allowed)
        if (s.charAt(0) == '0') {
            return false;
        }

        // Split in half and check if both halves are identical
        int mid = length / 2;
        String firstHalf = s.substring(0, mid);
        String secondHalf = s.substring(mid);

        return firstHalf.equals(secondHalf);
    }

    private static boolean isInvalidIdPart2(long num) {
        String s = Long.toString(num);
        int length = s.length();

        // Check if it starts with 0 (leading zeros not allowed)
        if (s.charAt(0) == '0') {
            return false;
        }

        // Try all possible pattern lengths from 1 to length/2
        for (int patternLength = 1; patternLength <= length / 2; patternLength++) {
            if (length % patternLength == 0) {
                String pattern = s.substring(0, patternLength);
                int repetitions = length / patternLength;

                // Check if repeating the pattern gives us the original string
                StringBuilder sb = new StringBuilder();
                for (int i = 0; i < repetitions; i++) {
                    sb.append(pattern);
                }

                if (sb.toString().equals(s)) {
                    return true;
                }
            }
        }

        return false;
    }

    private static long part1() {
        long total = 0;

        String[] parts = inputText.split(",");
        for (String part : parts) {
            part = part.trim();
            if (part.contains("-")) {
                String[] rangeParts = part.split("-");
                if (rangeParts.length == 2) {
                    long start = Long.parseLong(rangeParts[0]);
                    long end = Long.parseLong(rangeParts[1]);

                    for (long num = start; num <= end; num++) {
                        if (isInvalidIdPart1(num)) {
                            total += num;
                        }
                    }
                }
            }
        }

        return total;
    }

    private static long part2() {
        long total = 0;

        String[] parts = inputText.split(",");
        for (String part : parts) {
            part = part.trim();
            if (part.contains("-")) {
                String[] rangeParts = part.split("-");
                if (rangeParts.length == 2) {
                    long start = Long.parseLong(rangeParts[0]);
                    long end = Long.parseLong(rangeParts[1]);

                    for (long num = start; num <= end; num++) {
                        if (isInvalidIdPart2(num)) {
                            total += num;
                        }
                    }
                }
            }
        }

        return total;
    }
}
